Default values for logical types don't get converted

I have an Avro schema with logical type decimal and default value set as byte zero. However this causes a problems for `org.apache.kafka.connect.data.SchemaBuilder` as when it tries to validate given default value against logical type it can not find the match as default value is byte array and expected type is `BigDecimal`. 

Here is the stacktrace:
```
ERROR 2017-12-15 15:01:27,470+0000 pool-5-thread-25 :: WorkerTask - Task is being killed and will not recover until manually restarted
ERROR 2017-12-15 15:01:27,471+0000 pool-5-thread-23 :: WorkerTask - Task pm-ci-kafka-to-s3 threw an uncaught and unrecoverable exception
org.apache.kafka.connect.errors.SchemaBuilderException: Invalid default value
    at org.apache.kafka.connect.data.SchemaBuilder.defaultValue(SchemaBuilder.java:129) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at io.confluent.connect.avro.AvroData.toConnectSchema(AvroData.java:1476) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at io.confluent.connect.avro.AvroData.toConnectSchema(AvroData.java:1359) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at io.confluent.connect.avro.AvroData.toConnectSchema(AvroData.java:1239) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at io.confluent.connect.avro.AvroData.toConnectSchema(AvroData.java:1348) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at io.confluent.connect.avro.AvroData.toConnectSchema(AvroData.java:1381) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at io.confluent.connect.avro.AvroData.toConnectSchema(AvroData.java:1359) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at io.confluent.connect.avro.AvroData.toConnectSchema(AvroData.java:1239) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at io.confluent.connect.avro.AvroData.toConnectData(AvroData.java:993) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at io.confluent.connect.avro.AvroConverter.toConnectData(AvroConverter.java:89) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at org.apache.kafka.connect.runtime.WorkerSinkTask.convertMessages(WorkerSinkTask.java:425) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:264) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:182) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:150) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:146) [kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:190) [kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_111]
    at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_111]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_111]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_111]
    at java.lang.Thread.run(Thread.java:745) [na:1.8.0_111]
Caused by: org.apache.kafka.connect.errors.DataException: Invalid Java object for schema type BYTES: class [B for field: "null"
    at org.apache.kafka.connect.data.ConnectSchema.validateValue(ConnectSchema.java:239) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at org.apache.kafka.connect.data.ConnectSchema.validateValue(ConnectSchema.java:209) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    at org.apache.kafka.connect.data.SchemaBuilder.defaultValue(SchemaBuilder.java:127) ~[kafka-connect-s3-0.2.25-jar-with-dependencies.jar:0.2.25]
    ... 20 common frames omitted
```

I see that potential solution for this can be to use logical converters and convert default values (eg. bytes) to logical types before passing them to `org.apache.kafka.connect.data.ConnectSchema.validateValue`.

Version used is 3.3.0 but I see that it might affect newer versions as well.

One of the workarounds I tried doing is having default value `null` but apparently Avro does not support `null` union for logical types :(

Here is the test I quickly came up with and it fails:
```
  @Test 
  public void testToConnectLogicalDecimal(){
    org.apache.avro.Schema avroFieldSchema = createDecimalSchema(true, 64);
    org.apache.avro.Schema avroSchema = org.apache.avro.SchemaBuilder.builder()
            .record("Record").fields().name("number").type(avroFieldSchema).withDefault(new byte[]{0}).endRecord();
    GenericRecord avroRecord = new GenericRecordBuilder(avroSchema).set("number", new BigDecimal(5)).build();
    avroData.toConnectData(avroSchema, avroRecord);
  }
```

Thanks,
Niksa Jakovljevic

--- a/avro-converter/src/main/java/io/confluent/connect/avro/AvroData.java
+++ b/avro-converter/src/main/java/io/confluent/connect/avro/AvroData.java
@@ -988,6 +988,10 @@ public class AvroData {
     Object defaultVal = null;
     if (fieldSchema.defaultValue() != null) {
       defaultVal = fieldSchema.defaultValue();
+
+      // If this is a logical, convert to the primitive form for the Avro default value
+      defaultVal = fromLogical(fieldSchema, defaultVal);
+
       // Avro doesn't handle a few types that Connect uses, so convert those explicitly here
       if (defaultVal instanceof Byte) {
         // byte are mapped to integers in Avro
@@ -1014,6 +1018,26 @@ public class AvroData {
     fields.add(field);
   }
 
+  private static Object fromLogical(Schema schema, Object value) {
+    if (schema != null && schema.name() != null) {
+      LogicalTypeConverter logicalConverter = TO_AVRO_LOGICAL_CONVERTERS.get(schema.name());
+      if (logicalConverter != null && value != null) {
+        return logicalConverter.convert(schema, value);
+      }
+    }
+    return value;
+  }
+
+  private static Object toLogical(Schema schema, Object value) {
+    if (schema != null && schema.name() != null) {
+      LogicalTypeConverter logicalConverter = TO_CONNECT_LOGICAL_CONVERTERS.get(schema.name());
+      if (logicalConverter != null && value != null) {
+        return logicalConverter.convert(schema, value);
+      }
+    }
+    return value;
+  }
+
   // Convert default values from Connect data format to Avro's format, which is an
   // org.codehaus.jackson.JsonNode. The default value is provided as an argument because even
   // though you can get a default value from the schema, default values for complex structures need
@@ -1023,13 +1047,7 @@ public class AvroData {
     try {
       // If this is a logical type, convert it from the convenient Java type to the underlying
       // serializeable format
-      Object defaultVal = value;
-      if (schema != null && schema.name() != null) {
-        LogicalTypeConverter logicalConverter = TO_AVRO_LOGICAL_CONVERTERS.get(schema.name());
-        if (logicalConverter != null && value != null) {
-          defaultVal = logicalConverter.convert(schema, value);
-        }
-      }
+      Object defaultVal = fromLogical(schema, value);
 
       switch (schema.type()) {
         case INT8:
@@ -1493,13 +1511,11 @@ public class AvroData {
                   + " property must be a JSON Integer."
                   + " https://avro.apache.org/docs/1.7.7/spec.html#Decimal");
             }
+            // Capture the precision as a parameter only if it is not the default
             Integer precision = precisionNode.asInt();
-            builder.parameter(CONNECT_AVRO_DECIMAL_PRECISION_PROP, precision.toString());
-          } else {
-            builder.parameter(
-                CONNECT_AVRO_DECIMAL_PRECISION_PROP,
-                CONNECT_AVRO_DECIMAL_PRECISION_DEFAULT.toString()
-            );
+            if (precision != CONNECT_AVRO_DECIMAL_PRECISION_DEFAULT) {
+              builder.parameter(CONNECT_AVRO_DECIMAL_PRECISION_PROP, precision.toString());
+            }
           }
         } else {
           builder = SchemaBuilder.bytes();
@@ -1766,6 +1782,15 @@ public class AvroData {
   }
 
   private Object defaultValueFromAvro(Schema schema,
+      org.apache.avro.Schema avroSchema,
+      Object value,
+      ToConnectContext toConnectContext) {
+    Object result = defaultValueFromAvroWithoutLogical(schema, avroSchema, value, toConnectContext);
+    // If the schema is a logical type, convert the primitive Avro default into the logical form
+    return toLogical(schema, result);
+  }
+
+  private Object defaultValueFromAvroWithoutLogical(Schema schema,
                                       org.apache.avro.Schema avroSchema,
                                       Object value,
                                       ToConnectContext toConnectContext) {

--- a/avro-converter/src/main/java/io/confluent/connect/avro/AvroData.java
+++ b/avro-converter/src/main/java/io/confluent/connect/avro/AvroData.java
@@ -990,7 +990,7 @@ public class AvroData {
       defaultVal = fieldSchema.defaultValue();
 
       // If this is a logical, convert to the primitive form for the Avro default value
-      defaultVal = fromLogical(fieldSchema, defaultVal);
+      defaultVal = toAvroLogical(fieldSchema, defaultVal);
 
       // Avro doesn't handle a few types that Connect uses, so convert those explicitly here
       if (defaultVal instanceof Byte) {
@@ -1018,7 +1018,7 @@ public class AvroData {
     fields.add(field);
   }
 
-  private static Object fromLogical(Schema schema, Object value) {
+  private static Object toAvroLogical(Schema schema, Object value) {
     if (schema != null && schema.name() != null) {
       LogicalTypeConverter logicalConverter = TO_AVRO_LOGICAL_CONVERTERS.get(schema.name());
       if (logicalConverter != null && value != null) {
@@ -1028,7 +1028,7 @@ public class AvroData {
     return value;
   }
 
-  private static Object toLogical(Schema schema, Object value) {
+  private static Object toConnectLogical(Schema schema, Object value) {
     if (schema != null && schema.name() != null) {
       LogicalTypeConverter logicalConverter = TO_CONNECT_LOGICAL_CONVERTERS.get(schema.name());
       if (logicalConverter != null && value != null) {
@@ -1047,7 +1047,7 @@ public class AvroData {
     try {
       // If this is a logical type, convert it from the convenient Java type to the underlying
       // serializeable format
-      Object defaultVal = fromLogical(schema, value);
+      Object defaultVal = toAvroLogical(schema, value);
 
       switch (schema.type()) {
         case INT8:
@@ -1787,7 +1787,7 @@ public class AvroData {
       ToConnectContext toConnectContext) {
     Object result = defaultValueFromAvroWithoutLogical(schema, avroSchema, value, toConnectContext);
     // If the schema is a logical type, convert the primitive Avro default into the logical form
-    return toLogical(schema, result);
+    return toConnectLogical(schema, result);
   }
 
   private Object defaultValueFromAvroWithoutLogical(Schema schema,

--- a/avro-converter/src/main/java/io/confluent/connect/avro/AvroData.java
+++ b/avro-converter/src/main/java/io/confluent/connect/avro/AvroData.java
@@ -47,6 +47,7 @@ import org.codehaus.jackson.node.ObjectNode;
 import java.io.IOException;
 import java.math.BigDecimal;
 import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
@@ -1809,9 +1810,17 @@ public class AvroData {
       case BYTES:
       case FIXED:
         try {
-          return jsonValue.getBinaryValue();
+          byte[] bytes;
+          if (jsonValue.isTextual()) {
+            // Avro's JSON form may be a quoted string, so decode the binary value
+            String encoded = jsonValue.getTextValue();
+            bytes = encoded.getBytes(StandardCharsets.ISO_8859_1);
+          } else {
+            bytes = jsonValue.getBinaryValue();
+          }
+          return bytes == null ? null : ByteBuffer.wrap(bytes);
         } catch (IOException e) {
-          throw new DataException("Invalid binary data in default value");
+          throw new DataException("Invalid binary data in default value", e);
         }
 
       case ARRAY: {
@@ -1835,7 +1844,7 @@ public class AvroData {
         while (fieldIt.hasNext()) {
           Map.Entry<String, JsonNode> field = fieldIt.next();
           Object converted = defaultValueFromAvro(
-              schema, avroSchema.getElementType(), field.getValue(), toConnectContext);
+              schema.valueSchema(), avroSchema.getValueType(), field.getValue(), toConnectContext);
           result.put(field.getKey(), converted);
         }
         return result;

--- a/avro-converter/src/test/java/io/confluent/connect/avro/AvroDataTest.java
+++ b/avro-converter/src/test/java/io/confluent/connect/avro/AvroDataTest.java
@@ -296,10 +296,20 @@ public class AvroDataTest {
                                  .field("map", SchemaBuilder.map(Schema.STRING_SCHEMA, Schema.INT32_SCHEMA).defaultValue(Collections.singletonMap("field", 1)).build())
                                  .build();
     // leave the struct empty so that only defaults are used
-    Struct struct = new Struct(schema);
+    Struct struct = new Struct(schema)
+        .put("int8", (byte) 2)
+        .put("int16", (short) 12)
+        .put("int32", 12)
+        .put("int64", 12L)
+        .put("float32", 12.2f)
+        .put("float64", 12.2)
+        .put("boolean", true)
+        .put("string", "foo")
+        .put("bytes", ByteBuffer.wrap("foo".getBytes()))
+        .put("array", Arrays.asList("a", "b", "c"))
+        .put("map", Collections.singletonMap("field", 1))
 
-    Object convertedRecord = avroData.fromConnectData(schema, struct);
-
+    // Define the expected Avro schema
     org.apache.avro.Schema complexMapElementSchema =
         org.apache.avro.SchemaBuilder
             .record("MapEntry").namespace("io.confluent.connect.avro").fields()
@@ -388,19 +398,44 @@ public class AvroDataTest {
         .set("map", Collections.singletonMap("field", 1))
         .build();
 
-    org.apache.avro.generic.GenericRecord convertedAvroRecord = (org.apache.avro.generic
-        .GenericRecord) convertedRecord;
-    assertSchemaEquals(avroSchema, convertedAvroRecord.getSchema());
-    assertSchemaEquals(avroRecord.getSchema(), convertedAvroRecord.getSchema());
+    SchemaAndValue schemaAndValue = new SchemaAndValue(schema, struct);
+    schemaAndValue = convertToConnect(avroSchema, avroRecord, schemaAndValue);
+    schemaAndValue = convertToConnect(avroSchema, avroRecord, schemaAndValue);
+    schemaAndValue = convertToConnect(avroSchema, avroRecord, schemaAndValue);
+    assertNotNull(schemaAndValue);
+  }
+
+  protected SchemaAndValue convertToConnect(
+      org.apache.avro.Schema expectedAvroSchema,
+      org.apache.avro.generic.GenericRecord expectedAvroRecord,
+      SchemaAndValue connectSchemaAndValue
+  ) {
+    Object convertedRecord = avroData.fromConnectData(
+        connectSchemaAndValue.schema(),
+        connectSchemaAndValue.value()
+    );
+
+    org.apache.avro.generic.GenericRecord convertedAvroRecord =
+        (org.apache.avro.generic.GenericRecord) convertedRecord;
+    assertSchemaEquals(expectedAvroSchema, convertedAvroRecord.getSchema());
+    assertSchemaEquals(expectedAvroRecord.getSchema(), convertedAvroRecord.getSchema());
 
     // This doesn't work because the long field's default value is an integer
     //assertEquals(avroRecord, convertedRecord);
     // We've already checked the schemas, so we just need to check the record field values
-    for (org.apache.avro.Schema.Field field : avroSchema.getFields()) {
+    for (org.apache.avro.Schema.Field field : expectedAvroSchema.getFields()) {
       Object actual = convertedAvroRecord.get(field.name());
-      Object expected = avroRecord.get(field.name());
-      assertEquals(expected, actual);
+      Object expected = expectedAvroRecord.get(field.name());
+      assertValueEquals(expected, actual);
     }
+
+    SchemaAndValue schemaAndValue = avroData.toConnectData(
+        convertedAvroRecord.getSchema(),
+        convertedRecord
+    );
+    assertEquals(connectSchemaAndValue.schema(), schemaAndValue.schema());
+    assertEquals(connectSchemaAndValue.value(), schemaAndValue.value());
+    return schemaAndValue;
   }
 
   @Test
@@ -1985,11 +2020,18 @@ public class AvroDataTest {
         assertEquals(msg, expectedDouble, actualDouble, expectedDouble / 100.0d);
         break;
       default:
-        if (!expectedDef.equals(actualDef)) {
-          int x = 0;
-        }
         assertEquals(msg, expectedDef, actualDef);
         break;
     }
+  }
+
+  protected void assertValueEquals(Object expected, Object actual) {
+    if (actual instanceof byte[]) {
+      actual = ByteBuffer.wrap((byte[]) actual);
+    }
+    if (expected instanceof byte[]) {
+      expected = ByteBuffer.wrap((byte[]) expected);
+    }
+    assertEquals(expected, actual);
   }
 }

--- a/avro-converter/src/test/java/io/confluent/connect/avro/AvroDataTest.java
+++ b/avro-converter/src/test/java/io/confluent/connect/avro/AvroDataTest.java
@@ -34,6 +34,7 @@ import org.apache.kafka.connect.data.Struct;
 import org.apache.kafka.connect.data.Time;
 import org.apache.kafka.connect.data.Timestamp;
 import org.apache.kafka.connect.errors.DataException;
+import org.codehaus.jackson.JsonNode;
 import org.codehaus.jackson.node.ArrayNode;
 import org.codehaus.jackson.node.IntNode;
 import org.codehaus.jackson.node.JsonNodeFactory;
@@ -282,6 +283,15 @@ public class AvroDataTest {
   
   @Test
   public void testFromConnectComplexWithDefaults() {
+    int dateDefVal = 100;
+    int timeDefVal = 1000 * 60 * 60 * 2;
+    long tsDefVal = 1000 * 60 * 60 * 24 * 365 + 100;
+    java.util.Date dateDef = Date.toLogical(Date.SCHEMA, dateDefVal);
+    java.util.Date timeDef = Time.toLogical(Time.SCHEMA, timeDefVal);
+    java.util.Date tsDef = Timestamp.toLogical(Timestamp.SCHEMA, tsDefVal);
+    BigDecimal decimalDef = new BigDecimal(BigInteger.valueOf(314159L), 5);
+    byte[] decimalDefVal = decimalDef.unscaledValue().toByteArray();
+
     Schema schema = SchemaBuilder.struct()
                                  .field("int8", SchemaBuilder.int8().defaultValue((byte) 2).doc("int8 field").build())
                                  .field("int16", SchemaBuilder.int16().defaultValue((short)12).doc("int16 field").build())
@@ -294,6 +304,10 @@ public class AvroDataTest {
                                  .field("bytes", SchemaBuilder.bytes().defaultValue(ByteBuffer.wrap("foo".getBytes())).doc("bytes field").build())
                                  .field("array", SchemaBuilder.array(Schema.STRING_SCHEMA).defaultValue(Arrays.asList("a", "b", "c")).build())
                                  .field("map", SchemaBuilder.map(Schema.STRING_SCHEMA, Schema.INT32_SCHEMA).defaultValue(Collections.singletonMap("field", 1)).build())
+                                 .field("date", Date.builder().defaultValue(dateDef).doc("date field").build())
+                                 .field("time", Time.builder().defaultValue(timeDef).doc("time field").build())
+                                 .field("ts", Timestamp.builder().defaultValue(tsDef).doc("ts field").build())
+                                 .field("decimal", Decimal.builder(5).defaultValue(decimalDef).doc("decimal field").build())
                                  .build();
     // leave the struct empty so that only defaults are used
     Struct struct = new Struct(schema)
@@ -308,6 +322,10 @@ public class AvroDataTest {
         .put("bytes", ByteBuffer.wrap("foo".getBytes()))
         .put("array", Arrays.asList("a", "b", "c"))
         .put("map", Collections.singletonMap("field", 1))
+        .put("date", dateDef)
+        .put("time", timeDef)
+        .put("ts", tsDef)
+        .put("decimal", decimalDef);
 
     // Define the expected Avro schema
     org.apache.avro.Schema complexMapElementSchema =
@@ -347,6 +365,37 @@ public class AvroDataTest {
     bytesSchema.addProp("connect.doc", "bytes field");
     bytesSchema.addProp("connect.default", JsonNodeFactory.instance.binaryNode("foo".getBytes()));
 
+    org.apache.avro.Schema dateSchema = org.apache.avro.SchemaBuilder.builder().intType();
+    dateSchema.addProp("connect.doc", "date field");
+    dateSchema.addProp("connect.default", JsonNodeFactory.instance.numberNode(dateDefVal));
+    dateSchema.addProp(AvroData.CONNECT_NAME_PROP, Date.LOGICAL_NAME);
+    dateSchema.addProp(AvroData.CONNECT_VERSION_PROP, 1);
+    dateSchema.addProp(AvroData.AVRO_LOGICAL_TYPE_PROP, AvroData.AVRO_LOGICAL_DATE);
+
+    org.apache.avro.Schema timeSchema = org.apache.avro.SchemaBuilder.builder().intType();
+    timeSchema.addProp("connect.doc", "time field");
+    timeSchema.addProp("connect.default", JsonNodeFactory.instance.numberNode(timeDefVal));
+    timeSchema.addProp(AvroData.CONNECT_NAME_PROP, Time.LOGICAL_NAME);
+    timeSchema.addProp(AvroData.CONNECT_VERSION_PROP, 1);
+    timeSchema.addProp(AvroData.AVRO_LOGICAL_TYPE_PROP, AvroData.AVRO_LOGICAL_TIME_MILLIS);
+
+    org.apache.avro.Schema tsSchema = org.apache.avro.SchemaBuilder.builder().longType();
+    tsSchema.addProp("connect.doc", "ts field");
+    tsSchema.addProp("connect.default", JsonNodeFactory.instance.numberNode(tsDefVal));
+    tsSchema.addProp(AvroData.CONNECT_NAME_PROP, Timestamp.LOGICAL_NAME);
+    tsSchema.addProp(AvroData.CONNECT_VERSION_PROP, 1);
+    tsSchema.addProp(AvroData.AVRO_LOGICAL_TYPE_PROP, AvroData.AVRO_LOGICAL_TIMESTAMP_MILLIS);
+
+    org.apache.avro.Schema decimalSchema = org.apache.avro.SchemaBuilder.builder().bytesType();
+    decimalSchema.addProp("scale", 5);
+    decimalSchema.addProp("precision", 64);
+    decimalSchema.addProp("connect.doc", "decimal field");
+    decimalSchema.addProp(AvroData.CONNECT_VERSION_PROP, 1);
+    decimalSchema.addProp("connect.default", JsonNodeFactory.instance.binaryNode(decimalDefVal));
+    decimalSchema.addProp("connect.parameters", parameters("scale", "5"));
+    decimalSchema.addProp(AvroData.CONNECT_NAME_PROP, Decimal.LOGICAL_NAME);
+    decimalSchema.addProp(AvroData.AVRO_LOGICAL_TYPE_PROP, AvroData.AVRO_LOGICAL_DECIMAL);
+
     org.apache.avro.Schema arraySchema = org.apache.avro.SchemaBuilder.builder().array().items().stringType();
     ArrayNode arrayNode = JsonNodeFactory.instance.arrayNode();
     arrayNode.add("a");
@@ -382,6 +431,10 @@ public class AvroDataTest {
         .name("bytes").doc("bytes field").type(bytesSchema).withDefault(ByteBuffer.wrap("foo".getBytes()))
         .name("array").type(arraySchema).withDefault(Arrays.asList("a", "b", "c"))
         .name("map").type(mapSchema).withDefault(Collections.singletonMap("field", 1))
+        .name("date").doc("date field").type(dateSchema).withDefault(dateDefVal)
+        .name("time").doc("time field").type(timeSchema).withDefault(timeDefVal)
+        .name("ts").doc("ts field").type(tsSchema).withDefault(tsDefVal)
+        .name("decimal").doc("decimal field").type(decimalSchema).withDefault(ByteBuffer.wrap(decimalDefVal))
         .endRecord();
     org.apache.avro.generic.GenericRecord avroRecord
         = new org.apache.avro.generic.GenericRecordBuilder(avroSchema)
@@ -396,6 +449,10 @@ public class AvroDataTest {
         .set("bytes", ByteBuffer.wrap("foo".getBytes()))
         .set("array", Arrays.asList("a", "b", "c"))
         .set("map", Collections.singletonMap("field", 1))
+        .set("date", dateDefVal)
+        .set("time", timeDefVal)
+        .set("ts", tsDefVal)
+        .set("decimal", decimalDefVal)
         .build();
 
     SchemaAndValue schemaAndValue = new SchemaAndValue(schema, struct);
@@ -2034,4 +2091,17 @@ public class AvroDataTest {
     }
     assertEquals(expected, actual);
   }
+
+  private JsonNode parameters(String key1, String v1) {
+    return parametersFromConnect(Collections.singletonMap(key1, v1));
+  }
+
+  private JsonNode parametersFromConnect(Map<String, String> params) {
+    ObjectNode result = JsonNodeFactory.instance.objectNode();
+    for (Map.Entry<String, String> entry : params.entrySet()) {
+      result.put(entry.getKey(), entry.getValue());
+    }
+    return result;
+  }
+
 }

